### ISSUE TITLE
[Intermittent Bug] Network Adapter randomly disappears when rebooting

### ISSUE THREAD

**Issue by benjaminwestern** [AsahiLinux/linux#68]
Followed the installer yesterday. All worked fine
Booted back to macOS to ensure things were working as intended over there still
Booted back to Asahi to find ethernet device missing
`ifconfig` only showed `l0`  and `wlan0`
`lspci -v` results showed no Ethernet Controller only:
```
01:00.0 Network controller: Broadcom Inc. and subsidiaries BRCM4378 Wireless Network Adapter (rev 03)
        Subsystem: Apple Inc. Device 4378
        Device tree node: /sys/firmware/devicetree/base/soc/pcie@690000000/pci@0,0/network@0,0
        Flags: bus master, fast devsel, latency 0, IRQ 128, IOMMU group 4
        Memory at 6c0c00000 (64-bit, non-prefetchable) [size=64K]
        Memory at 6c0000000 (64-bit, non-prefetchable) [size=8M]
        Capabilities: <access denied>
        Kernel driver in use: brcmfmac
        Kernel modules: brcmfmac

01:00.1 Network controller: Broadcom Inc. and subsidiaries BRCM4378 Bluetooth Controller (rev 03)
        Subsystem: Apple Inc. Device 4378
        Device tree node: /sys/firmware/devicetree/base/soc/pcie@690000000/pci@0,0/network@0,1
        Flags: bus master, fast devsel, latency 0, IRQ 127, IOMMU group 4
        Memory at 6c0c10000 (64-bit, non-prefetchable) [size=32K]
        Memory at 6c0800000 (64-bit, non-prefetchable) [size=4M]
        Capabilities: <access denied>
        Kernel driver in use: hci_bcm4377
        Kernel modules: hci_bcm4377

```
`dmesg | grep atla` returned no results

-----------------------------------------------------------------------

Rebooted again
Ran `lspci -v` and now the following showed:
```
01:00.0 Network controller: Broadcom Inc. and subsidiaries BRCM4378 Wireless Network Adapter (rev 03)
        Subsystem: Apple Inc. Device 4378
        Device tree node: /sys/firmware/devicetree/base/soc/pcie@690000000/pci@0,0/network@0,0
        Flags: bus master, fast devsel, latency 0, IRQ 128, IOMMU group 4
        Memory at 6c0c00000 (64-bit, non-prefetchable) [size=64K]
        Memory at 6c0000000 (64-bit, non-prefetchable) [size=8M]
        Capabilities: <access denied>
        Kernel driver in use: brcmfmac
        Kernel modules: brcmfmac

01:00.1 Network controller: Broadcom Inc. and subsidiaries BRCM4378 Bluetooth Controller (rev 03)
        Subsystem: Apple Inc. Device 4378
        Device tree node: /sys/firmware/devicetree/base/soc/pcie@690000000/pci@0,0/network@0,1
        Flags: bus master, fast devsel, latency 0, IRQ 127, IOMMU group 4
        Memory at 6c0c10000 (64-bit, non-prefetchable) [size=32K]
        Memory at 6c0800000 (64-bit, non-prefetchable) [size=4M]
        Capabilities: <access denied>
        Kernel driver in use: hci_bcm4377
        Kernel modules: hci_bcm4377

...

03:00.0 Ethernet controller: Aquantia Corp. Device 04c0 (rev 03)
        Subsystem: Apple Inc. Device 0223
        Device tree node: /sys/firmware/devicetree/base/soc/pcie@690000000/pci@2,0/ethernet@0,0
        Flags: bus master, fast devsel, latency 0, IRQ 116, IOMMU group 5
        Memory at 6c0e00000 (64-bit, non-prefetchable) [size=512K]
        Memory at 6c0e80000 (64-bit, non-prefetchable) [size=4K]
        Memory at 6c1000000 (64-bit, non-prefetchable) [size=4M]
        Capabilities: <access denied>
        Kernel driver in use: atlantic
        Kernel modules: atlantic

```

Ran `dmesg | grep atla` and now the following showed:
```
[    2.478726] atlantic 0000:03:00.0: Adding to iommu group 5
[    2.478776] atlantic 0000:03:00.0: enabling device (0000 -> 0002)
[    2.528622] atlantic: Detect ATL2FW 103000d
[    2.531107] atlantic 0000:03:00.0 end0: renamed from eth0
[    2.571464] brcmfmac 0000:01:00.0: Direct firmware load for brcm/brcmfmac4378b1-pcie.apple,atlantisb-RASP-m-6.11-X3.bin failed with error -2
[    2.571598] brcmfmac 0000:01:00.0: Direct firmware load for brcm/brcmfmac4378b1-pcie.apple,atlantisb-RASP-m-6.11.bin failed with error -2
[    2.571646] brcmfmac 0000:01:00.0: Direct firmware load for brcm/brcmfmac4378b1-pcie.apple,atlantisb-RASP-m.bin failed with error -2
[    2.571656] brcmfmac 0000:01:00.0: Direct firmware load for brcm/brcmfmac4378b1-pcie.apple,atlantisb-RASP.bin failed with error -2
[    2.571667] brcmfmac 0000:01:00.0: Direct firmware load for brcm/brcmfmac4378b1-pcie.apple,atlantisb-X3.bin failed with error -2
[    9.889006] atlantic 0000:03:00.0 end0: atlantic: link change old 0 new 1000

```

If there is anything I can do to assist in the case of the missing PCI device after reboot bug outlined above. Please let me know!

**Comment by ingenieroariel** [comment]
This happened to me on a Mac Studio, a reboot to OSX and then back to Asahi made the issue disappear.

```Apr 26 11:46:46 decepti kernel: pcieport 0000:00:02.0: AER: enabled with IRQ 134
Apr 26 11:46:46 decepti kernel: OF: /soc@200000000/pcie@590000000: no iommu-map translation for id 0x18 on (null)
Apr 26 11:46:46 decepti kernel: pcieport 0000:00:03.0: enabling device (0000 -> 0002)
Apr 26 11:46:46 decepti kernel: pcieport 0000:00:03.0: PME: Signaling with IRQ 136
Apr 26 11:46:46 decepti kernel: pcieport 0000:00:03.0: AER: enabled with IRQ 136
Apr 26 11:46:46 decepti kernel: brcmfmac: brcmf_fw_alloc_request: using brcm/brcmfmac4387c2-pcie for chip BCM4387/7
Apr 26 11:46:46 decepti kernel: brcmfmac 0000:01:00.0: Direct firmware load for brcm/brcmfmac4387c2-pcie.apple,okinawa-TPHN-u-4.7-X0.bin failed with error -2
Apr 26 11:46:46 decepti kernel: brcmfmac 0000:01:00.0: Direct firmware load for brcm/brcmfmac4387c2-pcie.apple,okinawa-TPHN-u-4.7.bin failed with error -2
Apr 26 11:46:46 decepti kernel: brcmfmac 0000:01:00.0: Direct firmware load for brcm/brcmfmac4387c2-pcie.apple,okinawa-TPHN-u.bin failed with error -2
Apr 26 11:46:46 decepti kernel: brcmfmac 0000:01:00.0: Direct firmware load for brcm/brcmfmac4387c2-pcie.apple,okinawa-TPHN.bin failed with error -2
Apr 26 11:46:46 decepti kernel: brcmfmac 0000:01:00.0: Direct firmware load for brcm/brcmfmac4387c2-pcie.apple,okinawa-X0.bin failed with error -2
Apr 26 11:46:47 decepti kernel: sdhci-pci 0000:02:00.0: Adding to iommu group 5
Apr 26 11:46:47 decepti kernel: sdhci-pci 0000:02:00.0: SDHCI controller found [17a0:9755] (rev 1)
Apr 26 11:46:47 decepti kernel: sdhci-pci 0000:02:00.0: enabling device (0000 -> 0002)
Apr 26 11:46:47 decepti kernel: pci 0000:04:00.0: PME# does not work under D0, disabling it
Apr 26 11:46:47 decepti kernel: pci 0000:04:00.0: enabling device (0000 -> 0002)```

**Comment by jannau** [comment]
I believe I "fixed" this issue today in [asahi-6.10.9-1](https://github.com/AsahiLinux/linux/releases/tag/asahi-6.10.9-1). It shouldn't be have been reproducible for booting through u-boot since a long time but it still reproduced for direct Linux boots through m1n1. The relevant commits are 3a69f693ef76 and bcfb4fd381e2

Please reopen if it still is reproducible or was this year for Distro boots via u-boot.
